### ISSUE TITLE
Draft: Communication layer abstraction for expanding communication properties.

### ISSUE THREAD

**Issue by Jakio815** [lf-lang/reactor-c#281]
- [ ] Abstract socket layer

**Comment by Jakio815** [comment]
Hi, @erlingrj I would like some insights of this.
Prof. Kim suggested to use `void` pointers inside `lf_network_port` instead of using #ifdef #endifs, and type cast them by using `switch case`. However, I'm concerned that at the compile time, the types will all have to be declared, which means all additional header files to use TLS, SST, or DDS will need to be included even though it is not used. What do you think is a better idea? Using ifdefs or type casting?

Marten and Prof. Kim recommended you, because you handled the C code adding different target platforms.

**Comment by erlingrj** [comment]
My initial thought is to make C-style inheritance. It would look like this:

```
// The "abstract" base struct, contains any fields which are common to all: 
typedef struct {
  char * name;
  int a;
} lf_network_port_t;
 
// Low-level socket child struct. Should be defined in its own header file
typedef struct {
  lf_network_port_t base;
  int socket;
} lf_network_port_socket_t;

// TLS socket child struct. This should be defined in its own header file
typedef struct {
  lf_network_port_t base;
  SSL_CTX *ctx;
  SSL *ssl;
} lf_network_port_tls_t;

```

Since the first field of `lf_network_port_socket_t` is a `lf_network_port_t`, then we can cast any `lf_network_port_socket_t` to a `lf_network_port_t` (also true for pointers). 

I would then create a "virtual" API for the "abstract" `lf_network_port_t` which should be implemented by each child

```
lf_network_port_t * lf_network_port_create();
void lf_network_port_free(lf_network_port_t * port);
int lf_network_port_send(lf_network_port_t *port, msg_t msg, dst_t dsg);
...
```

This would be implemented in the source files for each of different types of ports. E.g. for the socket:

```
lf_network_port_t *lf_network_port_create() {
  // Dynamically allocate the child struct
  lf_network_port_socket_t * sock = malloc(sizeof(lf_network_port_socket_t));
  // Create a socket
  sock.socket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
  // Return an "abstract" pointer
  return sock.base;
}

```

The challenge here becomes the following: "Can you define a common abstract API for all the network types so that they all have the same interface to the LF codebase?" The different network types have different ways of being configured and such, but can all of that be hidden behind a single, network-agnostic API?

So, in a sense, I agree with Prof Kim that we should use typecasts, but we should make it elaborate with inheritance and only do the typecasting to e.g. TLS socket inside of tls_socket.c and only include TLS specific header files there. Here is a nice resource for object-oriented programming in C: https://www.state-machine.com/oop

**Comment by Jakio815** [comment]
I totally agree that the largest challenge is as what you mentioned. "Is it possible to have a single API for all the network types". I also think it may not be possible actually, because not only the sending and receiving APIs, there are many different initializing settings to do for each network type.
My answer is kind of a "yes" because I think we could also wrap up the initialization and managing additional structs together. Not 100% sure but I'll have to go for a try.

I got your idea of separating each type into different files. 

Such as
```
///  lf_network_port_tls.h
// TLS socket child struct. This should be defined in its own header file
typedef struct {
  lf_network_port_t base;
  SSL_CTX *ctx;
  SSL *ssl;
} lf_network_port_tls_t;

lf_network_port_t * lf_network_port_create();
void lf_network_port_free(lf_network_port_t * port);
int lf_network_port_send(lf_network_port_t *port, msg_t msg, dst_t dsg);
...

```
Checking how the platforms works will be great. Thanks!
